packages/overture-schema-validation/README.md

```diff
@@ -0,0 +1,340 @@
+    wikidata_id: Optional[str] = Field(None, description="Wikidata ID")
+
+    @field_validator("country")
+    @classmethod
+    def validate_country_code(cls, v):
+        if not re.match(r"^[A-Z]{2}$", v):
+            raise ValueError("Invalid ISO 3166-1 alpha-2 country code")
+        return v
+
+    @field_validator("language")
+    @classmethod
+    def validate_language_tag(cls, v):
+        if not re.match(r"^[a-z]{2,3}(-[A-Za-z]{2,8})*$", v):
+            raise ValueError("Invalid IETF BCP-47 language tag")
+        return v
+
+    @field_validator("categories")
+    @classmethod
+    def validate_unique_categories(cls, v):
+        if len(v) != len(set(v)):
+            raise ValueError("Categories must be unique")
+        return v
+
+    @field_validator("wikidata_id")
+    @classmethod
+    def validate_wikidata_format(cls, v):
+        if v is not None and not re.match(r"^Q\d+$", v):
+            raise ValueError("Invalid Wikidata identifier format")
+        return v
+```
+
+**After (using constraints):**
+
+```python
+class PlaceProperties(BaseModel):
+    country: Annotated[str, CountryCodeConstraint()] = Field(
+        ..., description="Country code"
+    )
+    language: Annotated[str, LanguageTagConstraint()] = Field(
+        ..., description="Language tag"
+    )
+    categories: Annotated[List[str], UniqueItemsConstraint()] = Field(
+        ..., description="Categories"
+    )
+    # Domain-specific constraints removed for incremental approach
+```
+
+#### Using Constraint-Based Validation
+
+For complex model-level validation, use the mixin-based constraint system:
+
+```python
+from overture.schema.validation.mixin import ConstraintValidatedModel, at_least_one_of
+
+@at_least_one_of("max_speed", "min_speed")
+class SpeedLimitRule(ConstraintValidatedModel, BaseModel):
+    max_speed: Optional[Speed] = None
+    min_speed: Optional[Speed] = None
+```
+
+##### ⚠️ CRITICAL: Inheritance Order Matters
+
+When using `ConstraintValidatedModel`, it **MUST** come first in the inheritance list:
+
+```python
+# ✅ CORRECT - ConstraintValidatedModel first
+class MyModel(ConstraintValidatedModel, BaseModel):
+    pass
+
+# ❌ WRONG - Will not generate JSON Schema metadata
+class MyModel(BaseModel, ConstraintValidatedModel):
+    pass
+```
+
+This is due to Python's Method Resolution Order (MRO). When `ConstraintValidatedModel` comes first,
+its `model_json_schema` method is called, which adds constraint metadata to the generated JSON
+Schema.
+
+## Error Messages
+
+Constraints provide detailed, consistent error messages:
+
+```python
+# Invalid country code
+ValidationError: 1 validation error for MyModel
+country
+  Invalid ISO 3166-1 alpha-2 country code: USA [type=value_error]
+
+# Mutual exclusion violation
+ValidationError: 1 validation error for BoundaryModel
+is_land, is_territorial
+  Fields is_land, is_territorial are mutually exclusive and cannot all be true [type=value_error]
```

Comment:
Is this custom? Could we have an option to produce more machine-friendly output?